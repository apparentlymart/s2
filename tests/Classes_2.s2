#-*-c++-*-

layerinfo "type" = "core";
layerinfo "majorversion" = 0;

class A { 
  function write(string s); 
}
class B extends A { }

function A::write(string s) {
  println "A wrote: " + $s;
}

function B::write(string s) {
  println "B wrote: " + $s;
}

function getA (bool subclass) : A
{
    var A v = new A;
    if ($subclass) {
	$v = new B;
    }
    return $v;
}

function main ()
{
  var B b = new B;
  var A a = new A;
  $a->write("A");
  $b->write("B");

  $a = $b;
  $a->write("B from A");

  var A fa;
  $fa = getA(false);
  $fa->write("should be A");
  $fa = getA(true);
  $fa->write("should be B");
}

