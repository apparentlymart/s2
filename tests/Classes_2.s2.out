A wrote: A
B wrote: B
B wrote: B from A
A wrote: should be A
B wrote: should be B
